lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java

```diff
@@ -101,4 +115,207 @@ public void leaderIsInitialisedForAllMethods() {
         verify(mockedRegistry).timer(any());
         verify(mockedTimer).update(anyLong(), any());
     }
+
+    @Test
+    public void noLeaderElectionDurationBeforeLeaderElection() {
+        assertThat(timelockService.calculateLastLeaderElectionDuration()).isNotPresent();
+
+        timelockService.startTransactions(startTransactionsRequest);
+        assertThat(timelockService.calculateLastLeaderElectionDuration()).isNotPresent();
+
+        timelockService.getCommitTimestamps(commitTimestampsRequest);
+        assertThat(timelockService.calculateLastLeaderElectionDuration()).isNotPresent();
+    }
+
+    /**
+     * [ A ]
+     *       [ A ]
+     *             [ B ]
+     *       <=========>
+     */
+    @Test
+    public void detectLeaderElectionWithFreshLeader() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(3L, 6L, LEADER_1),
+                ImmutableSingleCall.of(10L, 15L, LEADER_2));
+        assertExpectedDuration(Instant.ofEpochMilli(3L), Instant.ofEpochMilli(15L));
+    }
+
+    /**
+     * [ A ]
+     *           [ A ]
+     *       [     A     ]
+     *                     [ B ]
+     *           <=============>
+     */
+    @Test
+    public void detectLeaderElectionWithFreshLeaderOverlappingRequests() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(6L, 10L, LEADER_1),
+                ImmutableSingleCall.of(3L, 15L, LEADER_1),
+                ImmutableSingleCall.of(14L, 21L, LEADER_2));
+        assertExpectedDuration(Instant.ofEpochMilli(6L), Instant.ofEpochMilli(21L));
+    }
+
+    /**
+     * [ A ]
+     *       [ A ]
+     *             [ B ]
+     *                   [ B ]
+     *       <=========>
+     */
+    @Test
+    public void detectLeaderElectionWithTwoLongTermLeaders() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(3L, 6L, LEADER_1),
+                ImmutableSingleCall.of(10L, 15L, LEADER_2),
+                ImmutableSingleCall.of(21L, 28L, LEADER_2));
+        assertExpectedDuration(Instant.ofEpochMilli(3L), Instant.ofEpochMilli(15L));
+    }
+
+    /**
+     * [ A ]
+     *       [ A ]
+     *             [ B ]
+     *                   [ B ]
+     *                         [ C ]
+     *             <===============>
```

Comment:
21 is the start of the second B call, so presumably this arrow should represent that?